exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java

```diff
@@ -269,13 +283,54 @@ protected void doOnMatch(RelOptRuleCall call, Filter filterRel, Project projectR
         int recordCount = 0;
         int qualifiedCount = 0;
 
-        // Inner loop: within each batch iterate over the PartitionLocations
-        for(PartitionLocation part: partitions){
-          if(!output.getAccessor().isNull(recordCount) && output.getAccessor().get(recordCount) == 1){
-            newPartitions.add(part);
-            qualifiedCount++;
+        if (checkForSingle &&
+            partitions.get(0).isCompositePartition() /* apply single partition check only for composite partitions */) {
+          // Inner loop: within each batch iterate over the PartitionLocations
+          for (PartitionLocation part : partitions) {
+            assert part.isCompositePartition();
+            if(!output.getAccessor().isNull(recordCount) && output.getAccessor().get(recordCount) == 1) {
+              newPartitions.add(part);
+              if (isSinglePartition) { // only need to do this if we are already single partition
+                // compose the array of partition values for the directories that are referenced by filter:
+                // e.g suppose the dir hierarchy is year/quarter/month and the query is:
+                //     SELECT * FROM T WHERE dir0=2015 AND dir1 = 'Q1',
+                // then for 2015/Q1/Feb, this will have ['2015', 'Q1', null]
```

Comment:
Good point.  The path [2015, null, Jan]  should *not* qualify for the single partition optimization because in the general case there could be multiple 'Jan' subdirectories under the dir1 directory.  We would still use the metadata cache but at the level of the dir0, so ideally one should get cacheFileRoot=/tmp/t2/2015.   I have a fix for this and will update the PR.